sign.go

```diff
@@ -112,6 +86,46 @@ func (file *File) MarshalForSignature() ([]byte, error) {
 	return file.Marshal()
 }
 
+// Hash takes an input and a signature algorithm and returns its hashed value
+func Hash(input []byte, sigalg uint32) (output []byte, h crypto.Hash, err error) {
+	// hash the signature block using the appropriate algorithm
+	var md hash.Hash
+	switch sigalg {
+	case SigAlgRsaPkcs1Sha1:
+		md = sha1.New()
+		h = crypto.SHA1
+	case SigAlgEcdsaP256Sha256:
+		md = sha256.New()
+		h = crypto.SHA256
+	case SigAlgRsaPkcs1Sha384, SigAlgEcdsaP384Sha384:
+		md = sha512.New384()
+		h = crypto.SHA384
+	default:
+		return nil, h, fmt.Errorf("unsupported signature algorithm")
+	}
+	md.Write(input)
+	return md.Sum(nil), h, nil
+}
+
+// Sign signs digest with the private key, possibly using entropy from rand
+func Sign(key crypto.PrivateKey, rand io.Reader, digest []byte, h crypto.Hash) (sigData []byte, err error) {
+	// call the signer interface of the private key to sign the hash
+	sigData, err = key.(crypto.Signer).Sign(rand, digest, h)
```

Comment:
can the cast to crypto.Signer fail and if so do we need to handle that?

Reply:
I added an assertion that will return an error if Signer isn't implemented